EIPS/eip-eof-metadata.md

```diff
@@ -0,0 +1,88 @@
+
+```
+container := header, body
+header :=
+    magic, version,
+    kind_type, type_size,
+    kind_code, num_code_sections, code_size+,
+    [kind_container, num_container_sections, container_size+,]
+    [kind_metadata, metadata_size,]
+    kind_data, data_size,
+    terminator
+body := types_section, code_section+, container_section*, [metadata_section], data_section
+types_section := (inputs, outputs, max_stack_height)+
+```
+
+#### Header
+
+| name          | length  | value         | description                                                                             |
+| ------------- | ------- | ------------- | --------------------------------------------------------------------------------------- |
+| ...           | ...     | ...           | ...                                                                                     |
+| kind_metadata | 1 byte  | 0x05          | kind marker for metadata size section                                                   |
+| metadata_size | 2 bytes | 0x0001-0xFFFF | 16-bit unsigned big-endian integer denoting the length of the metadata section content  |
+| kind_data     | 1 byte  | 0x04          | kind marker for data size section                                                       |
+| data_size     | 2 bytes | 0x0000-0xFFFF | 16-bit unsigned big-endian integer denoting the length of the data section content (\*) |
+| terminator    | 1 byte  | 0x00          | marks the end of the header                                                             |
+
+#### Body
+
+| name             | length   | value | description                 |
+| ---------------- | -------- | ----- | --------------------------- |
+| ...              | ...      | ...   | ...                         |
+| metadata_section | variable | n/a   | arbitrary sequence of bytes |
+| data_section     | variable | n/a   | arbitrary sequence of bytes |
+
+The strucure and the encoding of the `metadata_section` is not defined by this EIP. It is left to the compilers, tooling, or the contract developers to define the encoding and the content. The current practice by the Solidity and Vyper compilers is to use CBOR encoding.
+
+## Rationale
+
+The `metadata_section` in the `body`, as well as the `kind_metadata` and `metadata_size` fields in the `header`, are OPTIONAL. The `data_section` can change in its size and content during deployment, therefore it needs to be REQUIRED, even if the data is empty. The `metadata_section` is not expected to change during the deployment.
+
+The reason for placing the `metadata_section` before the `data_section`, and assigning `kind_metadata` the value `0x05` (and not `0x04`) is to make it easier for the existing EOF tooling adapt the changes. Additionally, if the `metadata_section` was placed before the `data_section`, changes to the `data_section` in deploy time would cause the `metadata_section` to shift. By placing the `metadata_section` before, this could be mitigated.
```

Comment:
```suggestion
The reason for placing the `metadata_section` before the `data_section`, and assigning `kind_metadata` the value `0x05` (and not `0x04`) is to make it easier for the existing EOF tooling adapt the changes. Additionally, if the `metadata_section` was placed after the `data_section`, changes to the `data_section` in deploy time would cause the `metadata_section` to shift. By placing the `metadata_section` before, this could be mitigated.
```
small mistake here